configuration/src/agent/mod.rs

```diff
@@ -58,6 +75,10 @@ macro_rules! decl_config {
         }
     ) => {
         affix::paste! {
+            pub(self) trait EnvOverridablePrivate {
+                fn load_env_overrides_private(&mut self) {}
+            }
```

Comment:
Feels a bit awkward to implement this as no-op function for [$name Config]. Thinking that declaration of env overrides should happen in the `decl_config!` macro rather than in a separate one, as we want every agent config to be overridable.

Reply:
Think the best way to try going about this would be to have every type within an agent config struct implement something like `FromEnv`. Then, in `decl_config!`, we could iterate through all the struct in fields, look for their corresponding env vars using `from_env()`, then load them and modify `&mut self`. Note that for fields that are `HashSet<T>` or `HashMap<T>`, we can implement custom util functions that call `<T>::from_env()` in a loop.

This way, the custom work lies within loading the struct fields themselves but loading the whole struct from env is DRYed up using macro.

Reply:
Something like:
```
macro_rules! decl_config {
    (
        $name:ident {
            $($(#[$tags:meta])* $prop:ident: $type:ty,)*
        }
    ) => {
        ... existing [$name Config] struct declaration ...

        impl [$name Config] {
	fn load_env_overrides(&mut self) {
          $(
              {
                 let field = <$prop>::from_env();
                 self.$prop = field;
              }
           )*
	}
    }
}
```

Reply:
@ltchang2019 We're deserializing here so the only _correct_ way to do this is to extend `serde` -- then we can have arbitrarily complex configs and it will just work.

The way to do that would be to extend e.g., `envy` -- use it as a jumping off point, add it to an existing crate so we don't have the overhead of another (internal) dep. It's a pretty thin impl anyway. And then extend it to our use cases.

As I suggested elsewhere, this is at best an 'ok' solution. But it's correct and tested and (minus tests) it's ~100 LOC and easy enough to get rid of when we have time to do the above. 

Given that we need to solidify this interface let's use this (once we get input from @arnaud036) and then do it properly when we have more time for it.

Reply:
As discussed, this needs an RFC about where it should land when we refactor it. I'm on that